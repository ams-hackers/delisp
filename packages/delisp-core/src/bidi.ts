import { InvariantViolation, assertNever } from "./invariant";
import { Expression } from "./syntax";
import { Type, tString, tNumber } from "./types";
import { unify } from "./unify";
import { printType } from "./type-printer";
import { last } from "./utils";
import { Location } from "./input";
import { printHighlightedExpr } from "./error-report";

export interface Environment {
  [variable: string]: Type;
}

function equalTypes(t1: Type, t2: Type): boolean {
  // Quick and dirty equality definition for trying out.
  const result = unify(t1, t2, {});
  return (
    result.tag === "unify-success" &&
    Object.keys(result.substitution).length === 0
  );
}

function typeError(message: string, entity: { location: Location }) {
  return new Error(printHighlightedExpr(message, entity.location));
}

export function infer(expr: Expression, env: Environment): Type | null {
  switch (expr.node.tag) {
    case "string":
      return tString;
    case "number":
      return tNumber;
    case "variable-reference": {
      const variableName = expr.node.name;
      const variableType = env[variableName];
      if (!variableType) {
        throw typeError(`Unknown variable ${variableName}`, expr);
      }
      return variableType;
    }
    case "type-annotation": {
      const typ = expr.node.typeWithWildcards.instantiate();
      check(expr.node.value, typ, env);
      return typ;
    }
    case "function-call": {
      const fnType = infer(expr.node.fn, env);
      if (!fnType) {
        throw typeError(
          `Can't called a function with unknown type.`,
          expr.node.fn
        );
      }

      if (
        !(
          fnType.node.tag === "application" &&
          fnType.node.op.node.tag === "constant" &&
          fnType.node.op.node.name === "->"
        )
      ) {
        throw typeError(
          `Looks like you are trying to do a function call on a ${printType(
            fnType
          )}`,
          expr.node.fn
        );
      }

      const argTypes = fnType.node.args.slice(0, -1);
      const outType = last(fnType.node.args);
      if (!outType) {
        throw new InvariantViolation(`Function with no output type?`);
      }

      expr.node.args.forEach((arg, i) => {
        check(arg, argTypes[i], env);
      });

      return outType;
    }

    default:
      return null;
  }
}

export function check(expr: Expression, type: Type, env: Environment) {
  switch (expr.node.tag) {
    case "string":
    case "number":
    case "variable-reference":
    case "type-annotation":
    case "function-call": {
      const inferred = infer(expr, env);
      if (inferred) {
        if (!equalTypes(inferred, type)) {
          throw typeError(
            `Expected type ${printType(type, false)}, but found ${printType(
              inferred
            )}.`,
            expr
          );
        }
      }
      return;
    }
    case "function": {
      if (
        !(
          type.node.tag === "application" &&
          type.node.op.node.tag === "constant" &&
          type.node.op.node.name === "->"
        )
      ) {
        throw typeError(
          `Expected a function, but found ${printType(type)}`,
          expr
        );
      }

      const argTypes = type.node.args.slice(0, -1);
      const outType = last(type.node.args);
      if (!outType) {
        throw new InvariantViolation(`Function with no output type?`);
      }

      const args = expr.node.lambdaList.positionalArgs.map(i => i.name);

      if (argTypes.length !== args.length) {
        throw typeError(
          `Wrong number of arguments. Expected ${argTypes.length} but found ${
            args.length
          }`,
          expr.node.lambdaList
        );
      }

      const newEnv = args.reduce(
        (tmpEnv, a, idx) => ({
          ...tmpEnv,
          [a]: argTypes[idx]
        }),
        env
      );

      check(last(expr.node.body)!, outType, newEnv);
      return;
    }
    case "conditional":
    case "let-bindings":
    case "record":
    case "vector":
      throw typeError(
        `Cannot check this expression for type ${printType(type)}.`,
        expr
      );
    default:
      assertNever(expr.node);
  }
}
