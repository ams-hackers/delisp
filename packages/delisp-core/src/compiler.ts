import { ASExpr, ASExprSymbol, ASExprList } from "./syntax";
import * as recast from "recast";

const debug = require("debug")("delisp:compiler");

type JSAST = object;
interface Environment {
  [symbol: string]: { symbol: ASExprSymbol; jsVar: string };
}

function parseLambdaList(x: ASExpr): ASExprSymbol[] {
  if (x.type !== "list")
    throw new Error("First argument of lambda should be a list");

  if (!x.elements.every(param => param.type === "symbol"))
    throw new Error("First argument of lambda should be a list of symbols");

  return x.elements as ASExprSymbol[];
}

function compileLambda(args: ASExpr[], env: Environment): JSAST {
  if (args.length !== 2) throw new Error("Lambda needs exactly 2 arguments");

  const params = parseLambdaList(args[0]);

  const newEnv = params.reduce(
    (e, param, ix) => ({
      ...e,
      [param.name]: { symbol: param, jsVar: `p${ix}` }
    }),
    env
  );

  return {
    type: "ArrowFunctionExpression",
    params: params.map(symbol => ({
      type: "Identifier",
      name: newEnv[symbol.name].jsVar
    })),
    body: compile(args[1], newEnv)
  };
}

function compileDefinition(args: ASExpr[], env: Environment): JSAST {
  if (args.length !== 2) throw new Error("Define needs exactly 2 arguments");

  const variable = args[0];
  if (variable.type !== "symbol")
    throw new Error("First argument of define should be a symbol");

  return {
    type: "AssignmentExpression",
    operator: "=",
    left: {
      type: "MemberExpression",
      computed: true,
      object: {
        type: "Identifier",
        name: "env"
      },
      property: {
        type: "Literal",
        value: variable.name
      }
    },
    right: compile(args[1], env)
  };
}

function compileList(fn: ASExpr, args: ASExpr[], env: Environment): JSAST {
  if (fn.type === "symbol") {
    switch (fn.name) {
      case "lambda":
        return compileLambda(args, env);
    }
  }

  return {
    type: "CallExpression",
    callee: compile(fn, env),
    arguments: args.map(arg => compile(arg, env))
  };
}

function compileVariable(symbol: ASExprSymbol, env: Environment): JSAST {
  if (symbol.name in env) {
    return {
      type: "Identifier",
      name: env[symbol.name].jsVar
    };
  } else {
    return {
      type: "MemberExpression",
      computed: true,
      object: {
        type: "Identifier",
        name: "env"
      },
      property: {
        type: "Literal",
        value: symbol.name
      }
    };
  }
}

export function compile(syntax: ASExpr, env: Environment): JSAST {
  switch (syntax.type) {
    case "number":
      return {
        type: "NumericLiteral",
        value: syntax.value
      };
    case "symbol":
      return compileVariable(syntax, env);
    case "string":
      return {
        type: "Literal",
        value: syntax.value
      };
    case "list":
      const [fn, ...args] = syntax.elements;
      return compileList(fn, args, env);
  }
}

function isSpecialForm(syntax: ASExprList, name: string) {
  if (syntax.elements.length > 0) {
    const first = syntax.elements[0];
    return first.type === "symbol" && first.name === name;
  } else {
    return false;
  }
}

function compileTopLevel(syntax: ASExpr, env: Environment): JSAST {
  if (syntax.type === "list" && isSpecialForm(syntax, "define")) {
    return compileDefinition(syntax.elements.slice(1), env);
  } else {
    return compile(syntax, env);
  }
}

export function compileModule(syntax: ASExpr): JSAST {
  return {
    type: "File",
    program: {
      type: "Program",
      sourceType: "module",
      body: [
        {
          type: "ExpressionStatement",
          expression: compileTopLevel(syntax, {})
        }
      ]
    }
  };
}

export function compileToString(syntax: ASExpr): string {
  const ast = compileModule(syntax);
  debug("jsast:", ast);
  return recast.print(ast).code;
}
