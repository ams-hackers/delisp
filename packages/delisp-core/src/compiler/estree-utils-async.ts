import * as JS from "estree";

type MaybeExpression =
  | boolean
  | number
  | string
  | undefined
  | MaybeJSExpressionNode[]
  | MaybeJSExpressionNode;

interface MaybeJSExpressionNode {
  type?: JS.Expression["type"];
  [key: string]: MaybeExpression;
}

/** Check if a body of JS expressions contain a await expression in the same scope.
 *
 * That is to say, not inside another function!
 */
export function isBodyAsync(body: JS.Expression[]) {
  function containsAwait(expr: MaybeExpression): boolean {
    if (
      !expr ||
      typeof expr === "string" ||
      typeof expr === "number" ||
      typeof expr === "boolean"
    ) {
      return false;
    }

    if (Array.isArray(expr)) {
      return expr.some(containsAwait);
    }

    if (!expr.type) {
      return false;
    }

    switch (expr.type) {
      case "AwaitExpression":
        return true;
      case "FunctionExpression":
      case "ArrowFunctionExpression":
        return false;
      default:
        return Object.entries(expr).some(([_, value]) => containsAwait(value));
    }
  }
  return body.some((e) => containsAwait(e as MaybeExpression));
}
