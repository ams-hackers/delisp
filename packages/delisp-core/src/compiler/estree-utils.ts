import * as JS from "estree";
import { isValidJSIdentifierName } from "./jsvariable";

export function member(obj: JS.Expression, prop: string): JS.MemberExpression {
  const dotNotation = isValidJSIdentifierName(prop);
  return {
    type: "MemberExpression",
    computed: !dotNotation,
    object: obj,
    property: dotNotation ? identifier(prop) : literal(prop)
  };
}

export function methodCall(
  e: JS.Expression,
  method: string,
  args: JS.Expression[]
): JS.CallExpression {
  return {
    type: "CallExpression",
    callee: member(e, method),
    arguments: args
  };
}

/** Generate an arrow function in JS
 *
 * @description This will execute all the expressions, returning the
 * last one. */
export function arrowFunction(
  args: JS.Pattern[],
  body: JS.Expression[]
): JS.ArrowFunctionExpression {
  if (body.length === 0) {
    throw new Error(`Empty body`);
  }

  if (body.length === 1) {
    const expr = body[0];
    return {
      type: "ArrowFunctionExpression",
      params: args,
      body: expr,
      generator: false,
      expression: true,
      async: false
    };
  } else {
    const middle: JS.Statement[] = body.slice(0, -1).map(e => ({
      type: "ExpressionStatement",
      expression: e
    }));

    const returning: JS.Statement = {
      type: "ReturnStatement",
      argument: body[body.length - 1]
    };

    return {
      type: "ArrowFunctionExpression",
      params: args,
      body: {
        type: "BlockStatement",
        body: [...middle, returning]
      },
      generator: false,
      expression: false,
      async: false
    };
  }
}

export function literal(value: number | string | boolean | null): JS.Literal {
  return {
    type: "Literal",
    value
  };
}

export function identifier(name: string): JS.Identifier {
  return {
    type: "Identifier",
    name
  };
}

export function op1(
  operator: JS.UnaryOperator,
  argument: JS.Expression
): JS.UnaryExpression {
  return { type: "UnaryExpression", operator, prefix: true, argument };
}

export function op(
  operator: JS.BinaryOperator,
  left: JS.Expression,
  right: JS.Expression
): JS.BinaryExpression {
  return { type: "BinaryExpression", operator, left, right };
}

export function primitiveCall(
  name: string,
  ...args: JS.Expression[]
): JS.Expression {
  return {
    type: "CallExpression",
    callee: identifier(name),
    arguments: args
  };
}

interface Property {
  key: string;
  value: JS.Expression;
}

export function objectExpression(properties: Property[]): JS.ObjectExpression {
  return {
    type: "ObjectExpression",
    properties: properties.map(p => {
      const validIdentifier = isValidJSIdentifierName(p.key);
      return {
        type: "Property",
        kind: "init",
        method: false,
        computed: false,
        shorthand:
          validIdentifier &&
          p.value.type === "Identifier" &&
          p.value.name === p.key,
        key: validIdentifier ? identifier(p.key) : literal(p.key),
        value: p.value
      };
    })
  };
}

export function requireModule(name: string): JS.Expression {
  return {
    type: "CallExpression",
    callee: identifier("require"),
    arguments: [literal(name)]
  };
}

export function requireNames(names: string[], source: string): JS.Statement {
  return {
    type: "VariableDeclaration",
    kind: "const",
    declarations: [
      {
        type: "VariableDeclarator",
        id: {
          type: "ObjectPattern",
          properties: names.map(name => ({
            type: "Property",
            kind: "init",
            key: identifier(name),
            value: identifier(name),
            computed: false,
            method: false,
            shorthand: true
          }))
        },
        init: {
          type: "CallExpression",
          callee: identifier("require"),
          arguments: [literal(source)]
        }
      }
    ]
  };
}
