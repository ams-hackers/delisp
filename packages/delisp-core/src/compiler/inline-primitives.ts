import * as JS from "estree";
import { readType } from "../type-convert";
import { InvariantViolation } from "../invariant";
import { generateUniqueTVar } from "../type-generate";
import { type } from "../type-tag";
import { generalize, isFunctionType } from "../type-utils";
import * as T from "../types";
import { range } from "../utils";
import {
  arrowFunction,
  identifier,
  literal,
  member,
  methodCall,
  objectExpression,
  op1,
  op,
  primitiveCall
} from "./estree-utils";

type InlineHandler = (args: JS.Expression[]) => JS.Expression;

interface InlinePrim {
  type: T.TypeSchema;
  // This handler is invoked to generate the output JS when the
  // primitive is used in value position.
  valueHandler: InlineHandler;
  // This handler is invoked to generate the output JS when the
  // primitive is used in function position.
  funcHandler: InlineHandler;
}

interface MagicPrim {
  matches: (name: string) => boolean;
  createPrimitive: (name: string) => InlinePrim;
}
const inlinefuncs = new Map<string, InlinePrim>();
const magicfuncs: MagicPrim[] = [];

function createInlinePrimitive(
  name: string,
  type: T.TypeSchema,
  valueHandler: InlineHandler,
  funcHandler: InlineHandler
) {
  const prim: InlinePrim = { type, valueHandler, funcHandler };
  return inlinefuncs.set(name, prim);
}

function definePrimitiveValue(
  name: string,
  typespec: string,
  handle: InlineHandler
) {
  const type = readType(typespec);
  return createInlinePrimitive(name, type, handle, handle);
}

function defineInlinePrimitive(
  name: string,
  typespec: string,
  handle: InlineHandler
) {
  const type = readType(typespec);
  const arity = typeArity(type.mono);

  /* If the primitive is used in a value position, a wrapper function
     will be created so the inlined primitive can be used as a
     function. */
  const valueHandler: InlineHandler = () => {
    const identifiers = range(arity).map(i => identifier(`x${i}`));
    const identifiersAndContext = [identifier("*context*"), ...identifiers];
    return {
      type: "ArrowFunctionExpression",
      params: [identifier("values"), ...identifiersAndContext],
      body: handle(identifiersAndContext),
      expression: true
    };
  };

  createInlinePrimitive(name, type, valueHandler, handle);
}

function defineMagicPrimitive(
  matches: MagicPrim["matches"],
  createPrimitive: MagicPrim["createPrimitive"]
) {
  magicfuncs.push({ matches, createPrimitive });
}

export function isInlinePrimitive(name: string) {
  return inlinefuncs.has(name) || magicfuncs.some(f => f.matches(name));
}

export function findInlinePrimitive(name: string): InlinePrim {
  const prim = inlinefuncs.get(name);
  if (prim) {
    return prim;
  }

  const magicPrim = magicfuncs.find(f => f.matches(name));
  if (magicPrim) {
    return magicPrim.createPrimitive(name);
  }

  throw new InvariantViolation(
    `${name} is not an primitive inline function call`
  );
}

function typeArity(type: T.Type): number {
  if (isFunctionType(type)) {
    // A type signature has 3 more elements than the arity of the
    // function:
    //    (-> context arg1 arg2 effect output)
    //
    // - context
    // - effect
    // - output
    return type.node.args.length - 3;
  } else {
    return 0;
  }
}

//
// Primitives
//

defineInlinePrimitive("not", "(-> _ctx boolean _ boolean)", ([_ctx, x]) =>
  op1("!", x)
);

defineInlinePrimitive(
  "and",
  "(-> _ctx boolean boolean _ boolean)",
  ([_ctx, left, right]) => ({
    type: "LogicalExpression",
    operator: "&&",
    left,
    right
  })
);

defineInlinePrimitive(
  "or",
  "(-> _ctx boolean bolean _ boolean)",
  ([_ctx, left, right]) => ({
    type: "LogicalExpression",
    operator: "||",
    left,
    right
  })
);

definePrimitiveValue("none", "none", () => {
  return identifier("undefined");
});

defineInlinePrimitive(
  "print",
  "(-> _ctx string (effect console | _) none)",
  ([_ctx, ...args]) => {
    return methodCall(identifier("console"), "log", args);
  }
);

defineInlinePrimitive(
  "+",
  "(-> _ctx number number _ number)",
  ([_ctx, ...args]) => op("+", args[0], args[1])
);

defineInlinePrimitive(
  "-",
  "(-> _ctx number number _ number)",
  ([_ctx, ...args]) => op("-", args[0], args[1])
);

defineInlinePrimitive(
  "*",
  "(-> _ctx number number _ number)",
  ([_ctx, ...args]) => op("*", args[0], args[1])
);

defineInlinePrimitive(
  "map",
  "(-> _ctx (-> _ctx a e (values b | _)) [a] e [b])",
  ([ctx, fn, vec]) => {
    return methodCall(vec, "map", [primitiveCall("bindPrimaryValue", fn, ctx)]);
  }
);

defineInlinePrimitive(
  "filter",
  "(-> _ctx (-> _ctx a _ (values boolean | _)) [a] _ [a])",
  ([ctx, predicate, vec]) => {
    return methodCall(vec, "filter", [
      primitiveCall("bindPrimaryValue", predicate, ctx)
    ]);
  }
);

defineInlinePrimitive(
  "fold",
  "(-> _ctx (-> _ctx b a _ (values b | _)) [a] b _ b)",
  ([ctx, fn, vec, init]) => {
    return methodCall(vec, "reduce", [
      primitiveCall("bindPrimaryValue", fn, ctx),
      init
    ]);
  }
);

defineInlinePrimitive(
  "append",
  "(-> _ctx [a] [a] _ [a])",
  ([_ctx, vec1, vec2]) => {
    return methodCall(vec1, "concat", [vec2]);
  }
);

defineInlinePrimitive("reverse", "(-> _ctx [a] _ [a])", ([_ctx, vec]) => {
  return methodCall(methodCall(vec, "slice", []), "reverse", []);
});

defineInlinePrimitive("length", "(-> _ctx [a] _ number)", ([_ctx, vec]) =>
  member(vec, "length")
);

defineInlinePrimitive(
  "=",
  "(-> _ctx number number _ boolean)",
  ([_ctx, x, y]) => op("===", x, y)
);

defineInlinePrimitive("zero?", "(-> _ctx number _ boolean)", ([_ctx, x]) =>
  op("===", x, literal(0))
);

defineInlinePrimitive(
  "<=",
  "(-> _ctx number number _ boolean)",
  ([_ctx, x, y]) => op("<=", x, y)
);

defineInlinePrimitive(
  "<",
  "(-> _ctx number number _ boolean)",
  ([_ctx, x, y]) => op("<", x, y)
);

defineInlinePrimitive(
  ">",
  "(-> _ctx number number _ boolean)",
  ([_ctx, x, y]) => op(">", x, y)
);

defineInlinePrimitive(
  ">=",
  "(-> _ctx number number _ boolean)",
  ([_ctx, x, y]) => op(">=", x, y)
);

defineInlinePrimitive(
  "string=",
  "(-> _ctx string string _ boolean)",
  ([_ctx, x, y]) => op("===", x, y)
);

defineInlinePrimitive(
  "string-length",
  "(-> _ctx string _ number)",
  ([_ctx, str]) => member(str, "length")
);

defineInlinePrimitive(
  "string-upcase",
  "(-> _ctx string _ string)",
  ([_ctx, str]) => methodCall(str, "toUpperCase", [])
);

defineInlinePrimitive(
  "string-downcase",
  "(-> _ctx string _ string)",
  ([_ctx, str]) => methodCall(str, "toLowerCase", [])
);

defineInlinePrimitive(
  "string-append",
  "(-> _ctx string string _ string)",
  ([_ctx, str1, str2]) => op("+", str1, str2)
);

// Tuples

defineInlinePrimitive("pair", "(-> _ctx a b _ (* a b))", ([_ctx, a, b]) => {
  return primitiveCall("primPair", a, b);
});

defineInlinePrimitive("%fst", "(-> _ctx (* a b) _ a)", ([_ctx, pair]) => {
  return primitiveCall("primFst", pair);
});

defineInlinePrimitive("%snd", "(-> _ctx (* a b) _ b)", ([_ctx, pair]) => {
  return primitiveCall("primSnd", pair);
});

defineInlinePrimitive(
  "assert",
  "(-> _ctx boolean string (effect exp | _) none)",
  ([_ctx, x, msg]) => {
    return primitiveCall("assert", x, msg);
  }
);

//
// Vectors
//

defineInlinePrimitive("cons", "(-> _ctx a [a] _ [a])", ([_ctx, x, lst]) => ({
  type: "ArrayExpression",
  elements: [x, { type: "SpreadElement", argument: lst }]
}));

defineInlinePrimitive("empty?", "(-> _ctx [a] _ boolean)", ([_ctx, lst]) =>
  op("===", member(lst, "length"), literal(0))
);

/*
matches `:foo` and inlines the lens with type

  (-> a _ (values b (-> c d)))

*/
defineMagicPrimitive(
  name => name[0] === ":" && name.length > 1,
  name => {
    const fieldType = generateUniqueTVar();
    const newFieldType = generateUniqueTVar();

    const extendsType = generateUniqueTVar();
    const recordType = T.record(
      [{ label: name, type: fieldType }],
      extendsType
    );
    const newRecordType = T.record(
      [{ label: name, type: newFieldType }],
      extendsType
    );

    const lensType = generalize(
      type`(-> _ctx1 ${recordType}
               _
               (values ${fieldType}
                       (-> _ctx2 ${newFieldType} _ ${newRecordType})))`,
      []
    );

    const jsname = name.replace(/^:/, "");

    const handler = () =>
      arrowFunction(
        [identifier("values"), identifier("*context*"), identifier("obj")],
        [
          primitiveCall(
            "values",
            member(identifier("obj"), jsname),
            arrowFunction(
              [
                identifier("values"),
                identifier("*context*"),
                identifier("val")
              ],
              [
                methodCall(identifier("Object"), "assign", [
                  objectExpression([]),
                  identifier("obj"),
                  objectExpression([
                    {
                      key: jsname,
                      value: identifier("val")
                    }
                  ])
                ])
              ]
            )
          )
        ]
      );

    return {
      type: lensType,
      valueHandler: handler,
      funcHandler: args => ({
        type: "CallExpression",
        callee: handler(),
        arguments: [identifier("values"), ...args]
      })
    };
  }
);
