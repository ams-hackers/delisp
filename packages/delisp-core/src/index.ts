export * from "./compiler";
export * from "./syntax-convert";
export * from "./error-report";
export * from "./eval";
export * from "./infer";
export * from "./infer-environment";
export * from "./module";
export * from "./module-dependencies";
export * from "./prettier";
export * from "./primitives";
export * from "./printer";
export * from "./syntax";
export * from "./syntax-typed";
export * from "./syntax-scoped";
export * from "./syntax-utils";
export * from "./type-printer";
export * from "./type-utils";
export { Type } from "./types";
export * as T from "./types";
export * from "./typescript-generation";
export * from "./macroexpand";
