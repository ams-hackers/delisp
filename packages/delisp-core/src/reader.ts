//
// reader.ts --- The reader for Delisp
//

// import { List, Syntax } from "./syntax";

//
// S-Expression parser
//

import {
  Parser,
  regex,
  alternatives,
  many,
  delimited
} from "./parser-combinators";

import { ASExpr } from "./syntax";

const spaces = regex(/^\s*/);

function spaced<A>(x: Parser<A>) {
  return delimited(spaces, x, spaces);
}

const number: Parser<ASExpr> = regex(/[0-9]+/)
  .map(
    (string, location): ASExpr => ({
      type: "number",
      value: parseInt(string, 10),
      location
    })
  )
  .description("number");

const doubleQuote = regex(/"/).description("double quote");

const stringChar = regex(/[^"\\]/).description("non-escaped string character");
const stringEscapedChar = regex(/\\[n\\]/).description(
  "escaped string character"
);
const stringConsitutent = alternatives(stringChar, stringEscapedChar);

const string = doubleQuote
  .then(many(stringConsitutent))
  .skip(doubleQuote)
  .map(
    (chars, location): ASExpr => ({
      type: "string",
      value: chars.join(""),
      location
    })
  );

const symbol: Parser<ASExpr> = regex(/[a-zA-Z+<>!@$%^*/-]+/)
  .map(
    (name, location): ASExpr => ({
      type: "symbol",
      name,
      location
    })
  )
  .description("symbol");

const atom: Parser<ASExpr> = alternatives(number, string, symbol).description(
  "atom"
);

const leftParen = regex(/\(/).description("open parenthesis");
const rightParen = regex(/\)/).description("close parenthesis");

function list(x: Parser<ASExpr>): Parser<ASExpr> {
  return leftParen
    .then(many(x))
    .skip(spaces)
    .skip(rightParen)
    .map(
      (elements, location): ASExpr => ({
        type: "list",
        elements,
        location
      })
    );
}

const sexpr: Parser<ASExpr> = spaced(atom.or(() => list(sexpr)));

//
// Parser a Delisp expression from a string
//
export function readFromString(str: string) {
  return sexpr.parse(str);
}
