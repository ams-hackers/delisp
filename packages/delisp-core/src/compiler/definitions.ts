import * as JS from "estree";

import { identifier } from "./estree-utils";

export interface DefinitionBackend {
  define(name: string, value: JS.Expression): JS.Statement;
  access(name: string): JS.Expression;
}

//
// Handle static definitions. Those are compiled to static JS
// functions as we won't allow to redefine them. Used during batch
// compilation.
//
export const staticDefinition: DefinitionBackend = {
  define(name, value) {
    return {
      type: "VariableDeclaration",
      kind: "const",
      declarations: [
        {
          type: "VariableDeclarator",
          id: identifier(name),
          init: value
        }
      ]
    };
  },

  access(name) {
    return identifier(name);
  }
};

//
// Dynamic definitions are used in the REPL to allow redefinitions and
// introspection.
//
export function dynamicDefinition(container: string): DefinitionBackend {
  return {
    define(name, value) {
      return {
        type: "ExpressionStatement",
        expression: {
          type: "AssignmentExpression",
          operator: "=",
          left: {
            type: "MemberExpression",
            computed: true,
            object: identifier(container),
            property: {
              type: "Literal",
              value: name
            }
          },
          right: value
        }
      };
    },

    access(name) {
      return {
        type: "MemberExpression",
        computed: true,
        object: identifier(container),
        property: {
          type: "Literal",
          value: name
        }
      };
    }
  };
}
