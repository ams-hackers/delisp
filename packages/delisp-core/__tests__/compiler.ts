import { readFromString } from "../src/reader";
import { compileToString } from "../src/compiler";

describe("Compiler", () => {
  describe("Error messages", () => {
    function compileError(str: string): string {
      try {
        const ast = readFromString(str);
        compileToString(ast);
        throw new Error(`FATAL: EXPRESSION DID NOT FAIL TO COMPILE.`);
      } catch (err) {
        return err.message;
      }
    }

    it("generate nice error message for invalid lambda expressions", () => {
      expect(compileError("(lambda)")).toMatchSnapshot();
      expect(compileError("(lambda 7)")).toMatchSnapshot();
      expect(compileError("(lambda 7 5)")).toMatchSnapshot();
      expect(compileError("(lambda (7) x)")).toMatchSnapshot();
      expect(compileError("(lambda (x 7) x)")).toMatchSnapshot();
    });

    it("generate nice error message for invalid definitions", () => {
      expect(compileError("(define)")).toMatchSnapshot();
      expect(compileError("(define 5)")).toMatchSnapshot();
      expect(compileError("(define x)")).toMatchSnapshot();
      expect(compileError("(define 4 2)")).toMatchSnapshot();
    });
  });
});
