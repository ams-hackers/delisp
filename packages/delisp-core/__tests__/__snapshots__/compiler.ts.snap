// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler Error messages generate nice error message for invalid definitions 1`] = `
"file:1:7: 'define' needs exactly 2 arguments, got 0
(define)
-------^"
`;

exports[`Compiler Error messages generate nice error message for invalid definitions 2`] = `
"file:1:9: 'define' needs exactly 2 arguments, got 1
(define 5)
---------^"
`;

exports[`Compiler Error messages generate nice error message for invalid definitions 3`] = `
"file:1:9: 'define' needs exactly 2 arguments, got 1
(define x)
---------^"
`;

exports[`Compiler Error messages generate nice error message for invalid definitions 4`] = `
"file:1:8: 'define' expected a symbol
(define 4 2)
--------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 1`] = `
"file:1:7: 'lambda' needs exactly 2 arguments, got 0
(lambda)
-------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 2`] = `
"file:1:9: 'lambda' needs exactly 2 arguments, got 1
(lambda 7)
---------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 3`] = `
"file:1:8: Expected a list of arguments
(lambda 7 5)
--------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 4`] = `
"file:1:9: A list of arguments should be made of symbols
(lambda (7) x)
---------^"
`;

exports[`Compiler Error messages generate nice error message for invalid lambda expressions 5`] = `
"file:1:11: A list of arguments should be made of symbols
(lambda (x 7) x)
-----------^"
`;
